Simon-Initiative/oli-torus
# [MER-1097] - Automation test data setup

Here at Argos, we're developing a suite of automated tests using cypress to test lesson content as part of a post-deployment verification. This PR implements a new API that allows those tests to set up and tear down the data needed to run those tests. Since these are deployment-verification tests, we want the ability to run the tests against remote torus instances, including our production instance.

The easiest way to test this new API:
1. Create a new API key via the admin interface [Manage Third-Party API Keys](https://localhost:8043/admin/api_keys)
2. Make sure that key has `Automation Data Setup Enabled` enabled.
3. Take the generated key and base-64 encode it. Make note of this value, you won't see it again.
4. Go to the swagger page and pick the automated test data setup endpoint `/api/v1/docs#/Automated%20Test%20Data%20Setup%20Service/OliWeb.Api.AutomationSetupController.setup`
5. Click the lock icon and paste in the value from step 3
6. Click the Try It option
7. Select the api options and click execute. (You probably want all true values and pick a zip file of an exported project)
![image](https://user-images.githubusercontent.com/333265/168141017-9b945d28-a0df-49ca-9c4e-58d54dd09845.png)

You should then get back information that an automated test can use to test the uploaded content.

```
{
  "author": {
    "email": "author-e2e-test-1652379065703091800@argos.test",
    "id": 78,
    "password": "k7vcskew@ree$s1pfnzb"
  },
  "educator": {
    "email": "educator-e2e-test-1652379066542668000@argos.test",
    "id": 88,
    "password": "aum6jc$yosbmjl3bpjyc"
  },
  "learner": {
    "email": "learner-e2e-test-1652379067188244800@argos.test",
    "id": 89,
    "password": "r14bwc0drzill#gfq!_6"
  },
  "project": {
    "id": 44,
    "slug": "cc_balancing_chemical_reaction_5odio",
    "title": "CC: Balancing Chemical Reactions 95858"
  },
  "section": {
    "id": 25,
    "slug": "automation_test_section_sasp2"
  },
  "success": true
}
```

Once your tests have completed running, you can use the teardown api endpoint to delete the data that had been created. Since that endpoint deletes data, we tried to be as conservative as possible to avoid deleting any real non-test data, therefore you must supply email & password credentials of each account to delete, and we run a variety of checks against deleted resources to make sure they were initialized by the setup function.
